subsys/net/ip/net_pkt.c

```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
 	return clone;
 }
 
+/* New allocator and API starts here */
+
+#if defined(CONFIG_NET_BUF_FIXED_DATA_SIZE)
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout,
+					const char *caller, int line)
+#else
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout)
+#endif
+{
+	u32_t alloc_start = k_uptime_get_32();
+	struct net_buf *first = NULL;
+	struct net_buf *current = NULL;
+
+	while (size) {
+		struct net_buf *new;
+
+		new = net_buf_alloc_fixed(pool, timeout);
+		if (!new) {
+			goto error;
+		}
+
+		if (!first && !current) {
+			first = new;
+		} else {
+			current->frags = new;
+		}
+
+		current = new;
+		if (current->size > size) {
+			current->size = size;
+			size = 0;
+		} else {
+			size -= current->size;
+		}
+
+		if (timeout != K_NO_WAIT && timeout != K_FOREVER) {
+			u32_t diff = k_uptime_get_32() - alloc_start;
+
+			timeout -= min(timeout, diff);
+		}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+		NET_FRAG_CHECK_IF_NOT_IN_USE(new, new->ref + 1);
+
+		net_pkt_alloc_add(new, false, caller, line);
+
+		NET_DBG("%s (%s) [%d] frag %p ref %d (%s():%d)",
+			pool2str(pool), get_name(pool), get_frees(pool),
+			new, new->ref, caller, line);
+#endif
+	}
+
+	return first;
+error:
+	if (first) {
+		net_buf_unref(first);
+	}
+
+	return NULL;
+}
+
+#else /* !CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout,
+					const char *caller, int line)
+#else
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout)
+#endif
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc_len(pool, size, timeout);
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	NET_FRAG_CHECK_IF_NOT_IN_USE(buf, buf->ref + 1);
+
+	net_pkt_alloc_add(buf, false, caller, line);
+
+	NET_DBG("%s (%s) [%d] frag %p ref %d (%s():%d)",
+		pool2str(pool), get_name(pool), get_frees(pool),
+		buf, buf->ref, caller, line);
+#endif
+
+	return buf;
+}
+
+#endif /* CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+static size_t pkt_buffer_length(struct net_pkt *pkt,
+				size_t size,
+				enum net_ip_protocol proto,
+				size_t existing)
+{
+	size_t max_len = net_if_get_mtu(net_pkt_iface(pkt));
+	sa_family_t family = net_pkt_family(pkt);
+
+	/* Family vs iface MTU */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		max_len = max(max_len, NET_IPV6_MTU);
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		max_len = max(max_len, NET_IPV4_MTU);
+	}
+
+	max_len -= existing;
+	max_len = min(size, max_len);
+
+	return min(size, max_len);
+}
+
+static size_t pkt_headers_buffer_length(struct net_pkt *pkt,
```

Comment:
Log message below says this returns *estimation* of headers len, why that's not mirrored in the name?

Reply:
Sure pkt_estitame_headers_length() then